### PR TITLE
Make TimeoutOr a Result type

### PR DESCRIPTION
**Pull request by beckerhe**
Similar to `Canceled` and `NotFound`, `Timeout` now becomes a result type and `TimeoutOr<T>` an alias for `Result<T, Timeout>`.

This also adds a helper function `WhenValueOrTimeout` which is a wrapper around `orbit_base::WhenAny`. But `WhenAny` returns a generic `std::variant<T, Timeout>` so, there is some glue code which converts it into a `Result,<T, Timeout>`.

### CODE REVIEW

File: src/QtUtils/include/QtUtils/CreateTimeout.h [hunk 1]
```diff
@@ -39,27 +46,38 @@ struct Timeout {};
...
 [[nodiscard]] inline orbit_base::Future<Timeout> CreateTimeout(absl::Duration duration) {
   return CreateTimeout(absl::ToChronoMilliseconds(duration));
 }
 
-// This is a helper type alias that is useful in combination with WhenAny. It provides a nicer way
-// to write the value type of the future that is returned by WhenAny when given an arbitrary future
-// and a Future<Timeout>.
 template <typename T>
-using ValueOrTimeout = std::variant<T, Timeout>;
+using TimeoutOr = Result<T, Timeout>;
```
Comment (beckerhe) [inline src/QtUtils/include/QtUtils/CreateTimeout.h:55]: Renaming this to be consistent with the other result types.

File: src/QtUtils/CreateTimeoutTest.cpp [hunk 2]
```diff
@@ -37,6 +40,7 @@ TEST(CreateTimeout, TimeoutCompletesEventually) {
 
 TEST(CreateTimeout, ParallelTimeoutsDontDeadlock) {
   std::vector<orbit_base::Future<Timeout>> timeouts;
+  timeouts.reserve(10);
```
Comment (beckerhe) [inline src/QtUtils/CreateTimeoutTest.cpp:43]: Unrelated change - suggested by clang-tidy.

**Review by akopich — APPROVED** [review]
LGTM
